### ISSUE TITLE
use /proc/self/task/<tid>/maps instead of /proc/self/maps

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#2539]
For address queries, we should investigate using the copy of maps in the task folder to avoid synchronization costs.  We should measure the cost, and futher measure the general cost of queries: maintaining the allmem vector is generally a headache and causes problems on Android and WSL and if a kernel query is fast enough we could perhaps stop using allmem.

**Comment by egrimley** [comment]
On this Xeon running 3.13.0, reading either of the "maps" files (open + read + close) takes about 9.5 us, and an mmap followed by an munmap takes about 0.5 us.

What are the main problems with maintaining the allmem vector (which I assume is just the memory map)? There's definitely the problem of stack-like allocations growing without the runtime being told. (I suppose you could replace them with non-stack-like allocations and extend them in response to a SIGSEGV that the app never sees.) There's definitely a problem with native threads modifying the memory map. (I suppose you could ptrace them. And you could optimise the case when there are no native threads.) Is there also a problem with tools doing memory allocation behind the runtime's back?

**Comment by derekbruening** [comment]
Did you measure an app with many threads running a workload?  The cost of the main maps vs the task maps is synchronization with the other threads.

I thought that most file parsing in DR (esp clients) used mmap: I'm surprised the maps reading is not already using mmap.  Are you measuring reading from memory or only mmap+munmap with no actual memory touches?

There have been many, many instances of the allmem vector not matching the maps file due to kernel variations in things like how adjacent mappings are merged or not.  Every new kernel variant has allmem mismatches that have to be dealt with.  That's why the first comment here talks about Android and WSL.  A quick search for "get_memory_info mismatch" shows 16 issues that have hit it in some form: https://github.com/DynamoRIO/dynamorio/issues?utf8=%E2%9C%93&q=is%3Aissue%20%22get_memory_info%20mismatch%22%20
Emulating the kernel, tracking our own changes to the address space: it's all complex and fragile and if we can just query on demand everything is much simpler.

Xref #2115

**Comment by egrimley** [comment]
No, my experiments were very simple: single-threaded, and I didn't use the mapped memory, which is a mistake, probably, as updating the page tables may be the most expensive part.

It seems reasonable for the kernel to merge or not merge adjacent regions with the same permissions; DynamoRIO ought to be able to cope with that. As I see it, the only things that are difficult in principle are stacks that grow by themselves and native threads. All other changes to the memory map happen through the same three wrappers round three system calls: mmap, mremap, munmap. (OK, there may be two versions of mmap on some architectures.) The code that maintains the memory map could be a component that gets unit-tested so that bugs don't turn up in complex situations later.

**Comment by derekbruening** [comment]
That sounds like what we thought when we implemented it years ago...

You can see in the code in various places that we just don't trust allmem and go with a "from_os" query instead.  The sanity checks vs what is considered ground truth in the maps file are to this day still firing frequently on Android, WSL, and various true-Linux scenarios.

**Comment by egrimley** [comment]
It would be good to know why the code is so untrustworthy. Is it just buggy? Has it ever been unit-tested? The message "can happen if os combines entries in /proc/pid/maps" suggests the implementation was never finished because merging adjacent regions with the same permissions is rather basic.

Thinking about this over the last hour, I've remembered something that might be a real problem: I vaguely remember that in some circumstances you can mmap one region on top of another, so that when you munmap only the top allocation is removed and the lower allocation is revealed again. No ordinary program does that, but if you were to ask me what would happen if you were to mmap 0-4, mmap 2-6, then munmap 1-3, I'd have to admit that I don't know. (Perhaps it would depend on the type of allocation!) But it seems to me unlikely that weird cases like that are causing the warning message quoted above.

**Comment by derekbruening** [comment]
There were many issues.  I don't recall any coming from actual bugs in the basics of allmem add/remove code, so I don't unit tests would help anything.  As you can see in the code, it does simple merging, but some kernels would merge .bss with .data even though .bss isn't file-backed, or split a region up based on events we cannot see like copy-on-write.  I should clarify my original statement: it's not our code that's untrustworthy by itself, it's the variety of Linux kernels out there that we found to be untrustworthy in providing consistent maps file behavior.  Ignoring the maps file completely also seemed like a poor idea.

See https://github.com/DynamoRIO/dynamorio/issues/853#issuecomment-64846356

I would also suggest looking in the code comments, in the aforementioned other issues, and looking into the Android and WSL discrepancies to get a feel for these problems.

Xref #853
